File: packages/b2c-tooling-sdk/src/auth/oauth-pkce.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
import {createHash, randomBytes} from 'node:crypto';
import {createServer, type Server, type IncomingMessage, type ServerResponse} from 'node:http';
import type {Socket} from 'node:net';
import {URL} from 'node:url';
import type {AuthStrategy, AccessTokenResponse, DecodedJWT, FetchInit} from './types.js';
import {getLogger} from '../logging/logger.js';
import {decodeJWT} from './oauth.js';
import {DEFAULT_ACCOUNT_MANAGER_HOST} from '../defaults.js';
import {findAuthSession, saveAuthSession, type AuthSession} from './session-store.js';

const DEFAULT_LOCAL_PORT = 8080;

const ACCESS_TOKEN_CACHE: Map<string, AccessTokenResponse> = new Map();
const PENDING_AUTH: Map<string, Promise<AccessTokenResponse>> = new Map();

/**
 * Configuration for the OAuth Authorization Code + PKCE flow.
 */
export interface PkceOAuthConfig {
  clientId: string;
  scopes?: string[];
  accountManagerHost?: string;
  /** Local port for the redirect server (default 8080 or SFCC_OAUTH_LOCAL_PORT). */
  localPort?: number;
  /** Override redirect URI (default `http://localhost:${localPort}` or SFCC_REDIRECT_URI). */
  redirectUri?: string;
  /** Custom browser opener. Receives the authorization URL. */
  openBrowser?: (url: string) => Promise<void>;
  /**
   * Persist tokens (access + refresh) to disk between CLI invocations,
   * keyed by clientId. When a refresh token is available it is used to
   * silently obtain a new access token instead of opening the browser.
   * Defaults to `true`.
   */
  persistSession?: boolean;
}

function base64url(buf: Buffer): string {
  return buf.toString('base64').replace(/=+$/, '').replaceAll('+', '-').replaceAll('/', '_');
}

function generatePkcePair(): {verifier: string; challenge: string} {
  const verifier = base64url(randomBytes(32));
  const challenge = base64url(createHash('sha256').update(verifier).digest());
  return {verifier, challenge};
}

async function openBrowserDefault(url: string): Promise<void> {
  try {
    const open = await import('open');
    await open.default(url);
  } catch {
    getLogger().debug('Could not automatically open browser');
  }
}

/**
 * OAuth 2.0 Authorization Code Flow with PKCE.
 *
 * Used for public clients (no client secret). Replaces the legacy implicit flow,
 * which is deprecated for public clients per OAuth 2.1.
 *
 * Flow:
 * 1. Generate PKCE verifier + S256 challenge.
 * 2. Open browser to `/dwsso/oauth2/authorize?response_type=code&code_challenge=...`.
 * 3. Capture redirect with `?code=...` on a localhost listener.
 * 4. POST `grant_type=authorization_code` + `code_verifier` to `/dwsso/oauth2/access_token`.
 *
 * Tokens may include a refresh_token (depends on client registration in Account Manager).
 */
export class PkceOAuthStrategy implements AuthStrategy {
  private accountManagerHost: string;
  private localPort: number;
  private redirectUri: string;
  private persistSession: boolean;
  private _hasHadSuccess = false;
  private _refreshToken: string | null = null;
  private _sub = '';
  private _hydrated = false;

  constructor(private config: PkceOAuthConfig) {
    this.accountManagerHost = config.accountManagerHost || DEFAULT_ACCOUNT_MANAGER_HOST;
    this.localPort = config.localPort || parseInt(process.env.SFCC_OAUTH_LOCAL_PORT || '', 10) || DEFAULT_LOCAL_PORT;
    this.redirectUri = config.redirectUri || process.env.SFCC_REDIRECT_URI || `http://localhost:${this.localPort}`;
    this.persistSession = config.persistSession !== false;

    getLogger().debug(
      {
        clientId: this.config.clientId,
        accountManagerHost: this.accountManagerHost,
        port: this.localPort,
        redirectUri: this.redirectUri,
        persistSession: this.persistSession,
      },
      '[Auth] PkceOAuthStrategy initialized',
    );
  }

  /**
   * Load any persisted session for this clientId. Idempotent.
   */
  private hydrate(): void {
    if (!this.persistSession || this._hydrated) return;
    this._hydrated = true;
    try {
      const stored = findAuthSession(this.config.clientId);
      if (!stored || stored.flow !== 'pkce') return;
      this._sub = stored.sub ?? '';
      this._refreshToken = stored.refreshToken ?? null;
      if (!ACCESS_TOKEN_CACHE.has(this.config.clientId) && stored.accessToken) {
        const expires = stored.expiresAt ? new Date(stored.expiresAt) : new Date(0);
        ACCESS_TOKEN_CACHE.set(this.config.clientId, {
          accessToken: stored.accessToken,
          expires,
          scopes: stored.scopes ?? [],
        });
      }
      getLogger().debug(
        {clientId: this.config.clientId, sub: this._sub, hasRefresh: this._refreshToken !== null},
        '[Auth] Hydrated PKCE session from store',
      );
    } catch (error) {
      getLogger().debug({err: error}, '[Auth] PKCE store hydration failed');
    }
  }

  async fetch(url: string, init: FetchInit = {}): Promise<Response> {
    const logger = getLogger();
    const method = init.method || 'GET';

    const token = await this.getAccessToken();

    const headers = new Headers(init.headers);
    headers.set('Authorization', `Bearer ${token}`);
    headers.set('x-dw-client-id', this.config.clientId);

    let res = await fetch(url, {...init, headers} as RequestInit);
    logger.debug({method, url, status: res.status}, '[Auth] Response');

    if (res.status !== 401) {
      this._hasHadSuccess = true;
    }

    if (res.status === 401 && this._hasHadSuccess) {
      logger.debug('[Auth] Received 401, invalidating PKCE token and retrying');
      this.invalidateToken();
      const newToken = await this.getAccessToken();
      headers.set('Authorization', `Bearer ${newToken}`);
      res = await fetch(url, {...init, headers} as RequestInit);
      logger.debug({method, url, status: res.status}, '[Auth] Retry response');
    }

    return res;
  }

  async getAuthorizationHeader(): Promise<string> {
    const token = await this.getAccessToken();
    return `Bearer ${token}`;
  }

  async getJWT(): Promise<DecodedJWT> {
    const token = await this.getAccessToken();
    return decodeJWT(token);
  }

  async getTokenResponse(): Promise<AccessTokenResponse> {
    this.hydrate();
    const cached = ACCESS_TOKEN_CACHE.get(this.config.clientId);
    if (cached && this.isCachedTokenUsable(cached)) {
      return cached;
    }
    if (this._refreshToken) {
      const refreshed = await this.tryRefresh();
      if (refreshed) return refreshed;
    }
    const tokenResponse = await this.runFlow();
    ACCESS_TOKEN_CACHE.set(this.config.clientId, tokenResponse);
    return tokenResponse;
  }

  invalidateToken(): void {
    ACCESS_TOKEN_CACHE.delete(this.config.clientId);
    this._refreshToken = null;
  }

  private isCachedTokenUsable(cached: AccessTokenResponse): boolean {
    const requiredScopes = this.config.scopes || [];
    const hasAllScopes = requiredScopes.every((scope) => cached.scopes.includes(scope));
    return hasAllScopes && Date.now() <= cached.expires.getTime();
  }

  private async getAccessToken(): Promise<string> {
    this.hydrate();
    const clientId = this.config.clientId;
    const cached = ACCESS_TOKEN_CACHE.get(clientId);
    if (cached && this.isCachedTokenUsable(cached)) {
      return cached.accessToken;
    }
    if (cached) {
      ACCESS_TOKEN_CACHE.delete(clientId);
    }

    const pending = PENDING_AUTH.get(clientId);
    if (pending) {
      const tokenResponse = await pending;
      return tokenResponse.accessToken;
    }

    const authPromise = (async (): Promise<AccessTokenResponse> => {
      if (this._refreshToken) {
        const refreshed = await this.tryRefresh();
        if (refreshed) return refreshed;
      }
      return this.runFlow();
    })();
    PENDING_AUTH.set(clientId, authPromise);
    try {
      const tokenResponse = await authPromise;
      ACCESS_TOKEN_CACHE.set(clientId, tokenResponse);
      return tokenResponse.accessToken;
    } finally {
      PENDING_AUTH.delete(clientId);
    }
  }

  /**
   * Exchange a stored refresh_token for a new access token. Returns null if no
   * refresh token is available or the exchange fails (e.g. revoked / expired).
   * On failure, the refresh token is forgotten so the next request triggers
   * the browser flow.
   */
  private async tryRefresh(): Promise<AccessTokenResponse | null> {
    const logger = getLogger();
    if (!this._refreshToken) return null;

    const tokenUrl = `https://${this.accountManagerHost}/dwsso/oauth2/access_token`;
    const body = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: this._refreshToken,
      client_id: this.config.clientId,
    });
    if (this.config.scopes && this.config.scopes.length > 0) {
      body.set('scope', this.config.scopes.join(' '));
    }

    let response: Response;
    try {
      response = await fetch(tokenUrl, {
        method: 'POST',
        headers: {'Content-Type': 'application/x-www-form-urlencoded'},
        body: body.toString(),
      });
    } catch (error) {
      logger.debug({err: error}, '[Auth] PKCE refresh request failed');
      this._refreshToken = null;
      return null;
    }

    if (!response.ok) {
      const text = await response.text();
      logger.debug({status: response.status, body: text}, '[Auth] PKCE refresh failed; falling back to browser flow');
      this._refreshToken = null;
      return null;
    }

    let parsed: {
      access_token: string;
      refresh_token?: string;
      expires_in?: number;
      scope?: string;
    };
    try {
      parsed = (await response.json()) as typeof parsed;
    } catch (error) {
      logger.debug({err: error}, '[Auth] PKCE refresh returned non-JSON response');
      this._refreshToken = null;
      return null;
    }

    const expiresIn = parsed.expires_in ?? 0;
    const expires = new Date(Date.now() + expiresIn * 1000);
    const scopes = parsed.scope ? parsed.scope.split(' ') : (this.config.scopes ?? []);
    const tokenResponse: AccessTokenResponse = {
      accessToken: parsed.access_token,
      expires,
      scopes,
    };
    if (parsed.refresh_token) {
      this._refreshToken = parsed.refresh_token;
    }
    this.persistTokens(tokenResponse);
    logger.debug({clientId: this.config.clientId}, '[Auth] PKCE token refreshed silently');
    return tokenResponse;
  }

  /**
   * Persist the current token + refresh token. No-op when `persistSession`
   * is disabled. Updates `_sub` from the JWT when available so the persisted
   * record carries the authenticated user identity for diagnostics.
   */
  private persistTokens(tokenResponse: AccessTokenResponse): void {
    if (!this.persistSession) return;

    let sub = this._sub;
    try {
      const decoded = decodeJWT(tokenResponse.accessToken);
      if (typeof decoded.payload.sub === 'string' && decoded.payload.sub.length > 0) {
        sub = decoded.payload.sub;
      }
    } catch {
      // ignore — token may not be a JWT, fall back to existing _sub
    }
    this._sub = sub;
    const record: AuthSession = {
      clientId: this.config.clientId,
      flow: 'pkce',
      accessToken: tokenResponse.accessToken,
      refreshToken: this._refreshToken,
      sub,
      expiresAt: tokenResponse.expires.toISOString(),
      scopes: tokenResponse.scopes,
      accountManagerHost: this.accountManagerHost,
    };
    try {
      saveAuthSession(record);
    } catch (error) {
      getLogger().debug({err: error}, '[Auth] Failed to persist PKCE session');
    }
  }

  private async runFlow(): Promise<AccessTokenResponse> {
    const logger = getLogger();
    const {verifier, challenge} = generatePkcePair();
    const state = base64url(randomBytes(16));

    const params = new URLSearchParams({
      client_id: this.config.clientId,
      redirect_uri: this.redirectUri,
      response_type: 'code',
      code_challenge: challenge,
      code_challenge_method: 'S256',
      state,
    });
    if (this.config.scopes && this.config.scopes.length > 0) {
      params.set('scope', this.config.scopes.join(' '));
    }

    const authorizeUrl = `https://${this.accountManagerHost}/dwsso/oauth2/authorize?${params.toString()}`;

    logger.info({url: authorizeUrl}, `Login URL: ${authorizeUrl}`);
    logger.info('If the URL does not open automatically, copy/paste it into a browser on this machine.');

    if (this.config.openBrowser) {
      await this.config.openBrowser(authorizeUrl);
    } else {
      await openBrowserDefault(authorizeUrl);
    }

    const code = await this.waitForAuthCode(state);
    logger.debug({codePrefix: code.slice(0, 8)}, '[Auth] Got authorization code, exchanging for token');

    const tokenUrl = `https://${this.accountManagerHost}/dwsso/oauth2/access_token`;
    const tokenBody = new URLSearchParams({
      grant_type: 'authorization_code',
      code,
      redirect_uri: this.redirectUri,
      client_id: this.config.clientId,
      code_verifier: verifier,
    });

    const tokenRes = await fetch(tokenUrl, {
      method: 'POST',
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: tokenBody.toString(),
    });
    const rawText = await tokenRes.text();
    if (!tokenRes.ok) {
      throw new Error(`PKCE token exchange failed (${tokenRes.status}): ${rawText}`);
    }

    let parsed: {
      access_token: string;
      refresh_token?: string;
      expires_in?: number;
      scope?: string;
    };
    try {
      parsed = JSON.parse(rawText);
    } catch {
      throw new Error(`PKCE token exchange returned non-JSON response: ${rawText}`);
    }

    const expiresIn = parsed.expires_in ?? 0;
    const expires = new Date(Date.now() + expiresIn * 1000);
    const scopes = parsed.scope ? parsed.scope.split(' ') : (this.config.scopes ?? []);
    const tokenResponse: AccessTokenResponse = {
      accessToken: parsed.access_token,
      expires,
      scopes,
    };

    if (parsed.refresh_token) {
      this._refreshToken = parsed.refresh_token;
    }
    this.persistTokens(tokenResponse);

    return tokenResponse;
  }

  private waitForAuthCode(expectedState: string): Promise<string> {
    const logger = getLogger();
    return new Promise<string>((resolve, reject) => {
      const sockets: Set<Socket> = new Set();
      const cleanup = () => {
        setTimeout(() => {
          server.close();
          for (const socket of sockets) socket.destroy();
        }, 100);
      };

      const server: Server = createServer((req: IncomingMessage, res: ServerResponse) => {
        const requestUrl = new URL(req.url || '/', `http://localhost:${this.localPort}`);
        const code = requestUrl.searchParams.get('code');
        const state = requestUrl.searchParams.get('state') ?? '';
        const error = requestUrl.searchParams.get('error');
        const errorDescription = requestUrl.searchParams.get('error_description');

        if (error) {
          res.writeHead(500, {'Content-Type': 'text/plain'});
          res.end(`Authentication failed: ${errorDescription ?? error}`);
          reject(new Error(`OAuth error: ${errorDescription ?? error}`));
          cleanup();
          return;
        }

        if (!code) {
          res.writeHead(404, {'Content-Type': 'text/plain'});
          res.end('Waiting for authorization code...');
          return;
        }

        if (state !== expectedState) {
          res.writeHead(400, {'Content-Type': 'text/plain'});
          res.end('State mismatch.');
          reject(new Error('OAuth state mismatch — aborting'));
          cleanup();
          return;
        }

        res.writeHead(200, {'Content-Type': 'text/plain'});
        res.end('Authentication successful! You may close this browser window and return to your terminal.');
        resolve(code);
        cleanup();
      });

      server.on('connection', (socket) => {
        sockets.add(socket);
        socket.on('close', () => sockets.delete(socket));
      });

      server.listen(this.localPort, () => {
        logger.debug({port: this.localPort}, `[Auth] PKCE redirect server listening on port ${this.localPort}`);
        logger.info('Waiting for user to authenticate...');
      });

      server.on('error', (err) => {
        const hint =
          'code' in err && (err as NodeJS.ErrnoException).code === 'EADDRINUSE'
            ? ` Port ${this.localPort} is in use; set SFCC_OAUTH_LOCAL_PORT or pass localPort to use a different port.`
            : '';
        reject(new Error(`Failed to start OAuth redirect server: ${err.message}.${hint}`));
      });
    });
  }
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/auth/oauth_pkce.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""OAuth 2.0 Authorization Code flow with PKCE.

Mirrors ``src/auth/oauth-pkce.ts``. Used for public clients (no client secret);
replaces the legacy implicit flow deprecated for public clients under OAuth 2.1.

Flow:

1. Generate a PKCE verifier + S256 challenge.
2. Open the browser to ``/dwsso/oauth2/authorize?response_type=code&code_challenge=...``.
3. Capture the redirect carrying ``?code=...`` on a localhost listener.
4. POST ``grant_type=authorization_code`` + ``code_verifier`` to ``/dwsso/oauth2/access_token``.

Tokens may include a ``refresh_token`` (depends on Account Manager client registration).
"""

from __future__ import annotations

import asyncio
import base64
import hashlib
import json
import os
import secrets
from collections.abc import Awaitable, Callable
from datetime import datetime, timedelta, timezone
from typing import Any
from urllib.parse import urlencode

import httpx

from b2c_tooling_sdk.auth.browser import open_browser_default
from b2c_tooling_sdk.auth.callback_server import CallbackResponse, run_callback_server
from b2c_tooling_sdk.auth.dispatch_fetch import dispatch_fetch
from b2c_tooling_sdk.auth.jwt_utils import decode_jwt
from b2c_tooling_sdk.auth.session_store import AuthSession, find_auth_session, save_auth_session
from b2c_tooling_sdk.auth.types import AccessTokenResponse, DecodedJWT
from b2c_tooling_sdk.defaults import DEFAULT_ACCOUNT_MANAGER_HOST, DEFAULT_LOCAL_PORT
from b2c_tooling_sdk.logging import get_logger

# Module-level caches so multiple instances sharing a clientId coordinate, matching
# the TS ACCESS_TOKEN_CACHE / PENDING_AUTH maps in oauth-pkce.ts.
_ACCESS_TOKEN_CACHE: dict[str, AccessTokenResponse] = {}
_PENDING_AUTH: dict[str, asyncio.Future[AccessTokenResponse]] = {}

# OAuth `error` codes that genuinely indicate the client is not registered as a
# PKCE-capable public client — the only failures the implicit fallback can rescue.
_PKCE_GRANT_UNSUPPORTED_OAUTH_ERRORS = frozenset(
    {
        "invalid_client",
        "unauthorized_client",
        "unsupported_response_type",
        "unsupported_grant_type",
    }
)


def reset_pkce_cache_for_testing() -> None:
    """Clear the module-level PKCE token / pending-auth caches (tests only)."""
    _ACCESS_TOKEN_CACHE.clear()
    _PENDING_AUTH.clear()


class PkceGrantUnsupportedError(Exception):
    """Raised when the Authorization Code + PKCE flow fails because the client is
    not registered for that grant (e.g. a legacy implicit-only public client or a
    missing/mismatched redirect URI) rather than a transient or user-driven failure.

    :class:`~b2c_tooling_sdk.auth.oauth_pkce_fallback.PkceWithImplicitFallbackStrategy`
    keys its automatic fallback off this type so it retries with the legacy implicit
    flow ONLY for grant/registration failures — never for user-cancel, state
    mismatch, or a port-in-use error.
    """

    def __init__(self, message: str, stage: str, oauth_error: str | None = None) -> None:
        super().__init__(message)
        self.stage = stage
        self.oauth_error = oauth_error


def _is_pkce_grant_unsupported_error(oauth_error: str | None) -> bool:
    """True only when an OAuth ``error`` code means the client cannot use the code grant."""
    return oauth_error is not None and oauth_error in _PKCE_GRANT_UNSUPPORTED_OAUTH_ERRORS


def _base64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _generate_pkce_pair() -> tuple[str, str]:
    verifier = _base64url(secrets.token_bytes(32))
    challenge = _base64url(hashlib.sha256(verifier.encode("ascii")).digest())
    return verifier, challenge


def _parse_oauth_error_body(text: str) -> tuple[str | None, str | None]:
    """Parse ``{error, error_description}`` from a token error body; ``(None, None)`` if not JSON."""
    try:
        parsed = json.loads(text)
    except Exception:  # noqa: BLE001 - non-JSON error body
        return None, None
    if not isinstance(parsed, dict):
        return None, None
    error = parsed.get("error")
    description = parsed.get("error_description")
    return (error if isinstance(error, str) else None, description if isinstance(description, str) else None)


class PkceOAuthConfig:
    """Configuration for the OAuth Authorization Code + PKCE flow."""

    def __init__(
        self,
        client_id: str,
        scopes: list[str] | None = None,
        account_manager_host: str | None = None,
        local_port: int | None = None,
        redirect_uri: str | None = None,
        open_browser: Callable[[str], Awaitable[None]] | None = None,
        persist_session: bool = True,
    ) -> None:
        self.client_id = client_id
        self.scopes = scopes
        self.account_manager_host = account_manager_host
        self.local_port = local_port
        self.redirect_uri = redirect_uri
        self.open_browser = open_browser
        self.persist_session = persist_session


class PkceOAuthStrategy:
    """OAuth 2.0 Authorization Code Flow with PKCE (public clients)."""

    auth_method = "user"

    def __init__(self, config: PkceOAuthConfig) -> None:
        self._config = config
        self._logger = get_logger("auth.oauth_pkce")
        self._account_manager_host = config.account_manager_host or DEFAULT_ACCOUNT_MANAGER_HOST
        env_port = os.environ.get("SFCC_OAUTH_LOCAL_PORT")
        self._local_port = config.local_port or _parse_int(env_port) or DEFAULT_LOCAL_PORT
        self._redirect_uri = (
            config.redirect_uri or os.environ.get("SFCC_REDIRECT_URI") or f"http://localhost:{self._local_port}"
        )
        self._persist_session = config.persist_session
        self._has_had_success = False
        self._refresh_token: str | None = None
        self._sub = ""
        self._hydrated = False

    def _hydrate(self) -> None:
        """Load any persisted PKCE session for this clientId. Idempotent."""
        if not self._persist_session or self._hydrated:
            return
        self._hydrated = True
        try:
            stored = find_auth_session(self._config.client_id)
            if stored is None or stored.flow != "pkce":
                return
            self._sub = stored.sub or ""
            self._refresh_token = stored.refresh_token
            if self._config.client_id not in _ACCESS_TOKEN_CACHE and stored.access_token:
                expires = (
                    _parse_iso(stored.expires_at) if stored.expires_at else datetime.fromtimestamp(0, tz=timezone.utc)
                )
                _ACCESS_TOKEN_CACHE[self._config.client_id] = AccessTokenResponse(
                    access_token=stored.access_token,
                    expires=expires,
                    scopes=stored.scopes or [],
                )
        except Exception as error:  # noqa: BLE001 - hydration is best-effort
            self._logger.debug("PKCE store hydration failed: %s", error)

    async def fetch(
        self,
        url: str,
        *,
        method: str = "GET",
        headers: dict[str, str] | None = None,
        content: Any = None,
        dispatcher: httpx.AsyncBaseTransport | None = None,
        **kwargs: Any,
    ) -> httpx.Response:
        """Perform a request with PKCE auth, retrying once on a post-success 401."""
        token = await self._get_access_token()
        request_headers = dict(headers or {})
        request_headers["Authorization"] = f"Bearer {token}"
        request_headers["x-dw-client-id"] = self._config.client_id

        response = await dispatch_fetch(
            url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
        )
        if response.status_code != 401:
            self._has_had_success = True
        if response.status_code == 401 and self._has_had_success:
            self.invalidate_token()
            new_token = await self._get_access_token()
            request_headers["Authorization"] = f"Bearer {new_token}"
            response = await dispatch_fetch(
                url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
            )
        return response

    async def get_authorization_header(self) -> str:
        """Return the ``Authorization`` header value (``Bearer <token>``)."""
        token = await self._get_access_token()
        return f"Bearer {token}"

    async def get_jwt(self) -> DecodedJWT:
        """Return the decoded (unverified) access-token JWT."""
        token = await self._get_access_token()
        return decode_jwt(token)

    async def get_token_response(self) -> AccessTokenResponse:
        """Return the full token response, refreshing or running the browser flow as needed."""
        self._hydrate()
        cached = _ACCESS_TOKEN_CACHE.get(self._config.client_id)
        if cached is not None and self._is_cached_token_usable(cached):
            return cached
        if self._refresh_token:
            refreshed = await self._try_refresh()
            if refreshed is not None:
                return refreshed
        token_response = await self._run_flow()
        _ACCESS_TOKEN_CACHE[self._config.client_id] = token_response
        return token_response

    def invalidate_token(self) -> None:
        """Drop only the cached access token; the refresh token is preserved for silent renewal."""
        _ACCESS_TOKEN_CACHE.pop(self._config.client_id, None)

    def _is_cached_token_usable(self, cached: AccessTokenResponse) -> bool:
        required_scopes = self._config.scopes or []
        has_all = all(scope in cached.scopes for scope in required_scopes)
        return has_all and _now() <= cached.expires

    async def _get_access_token(self) -> str:
        self._hydrate()
        client_id = self._config.client_id
        cached = _ACCESS_TOKEN_CACHE.get(client_id)
        if cached is not None and self._is_cached_token_usable(cached):
            return cached.access_token
        if cached is not None:
            _ACCESS_TOKEN_CACHE.pop(client_id, None)

        pending = _PENDING_AUTH.get(client_id)
        if pending is not None:
            token_response = await pending
            return token_response.access_token

        async def _auth() -> AccessTokenResponse:
            if self._refresh_token:
                refreshed = await self._try_refresh()
                if refreshed is not None:
                    return refreshed
            return await self._run_flow()

        auth_future = asyncio.ensure_future(_auth())
        _PENDING_AUTH[client_id] = auth_future
        try:
            token_response = await auth_future
            _ACCESS_TOKEN_CACHE[client_id] = token_response
            return token_response.access_token
        finally:
            _PENDING_AUTH.pop(client_id, None)

    async def _try_refresh(self) -> AccessTokenResponse | None:
        """Exchange a stored refresh token for a new access token. ``None`` on failure."""
        if not self._refresh_token:
            return None
        token_url = f"https://{self._account_manager_host}/dwsso/oauth2/access_token"
        params = {
            "grant_type": "refresh_token",
            "refresh_token": self._refresh_token,
            "client_id": self._config.client_id,
        }
        if self._config.scopes:
            params["scope"] = " ".join(self._config.scopes)
        body = _form_encode(params)

        try:
            response = await dispatch_fetch(
                token_url,
                method="POST",
                headers={"Content-Type": "application/x-www-form-urlencoded"},
                content=body,
            )
        except Exception as error:  # noqa: BLE001 - transport failure forgets refresh token
            self._logger.debug("PKCE refresh request failed: %s", error)
            self._refresh_token = None
            return None

        if not response.is_success:
            oauth_error, _ = _parse_oauth_error_body(response.text)
            self._logger.debug("PKCE refresh failed (%s); falling back to browser flow", oauth_error)
            self._refresh_token = None
            return None

        try:
            parsed = response.json()
        except Exception as error:  # noqa: BLE001 - non-JSON refresh response
            self._logger.debug("PKCE refresh returned non-JSON response: %s", error)
            self._refresh_token = None
            return None

        access_token = parsed.get("access_token")
        if not isinstance(access_token, str) or not access_token:
            self._logger.debug("PKCE refresh response did not contain an access token")
            self._refresh_token = None
            return None

        token_response = self._build_token_response(parsed)
        new_refresh = parsed.get("refresh_token")
        if isinstance(new_refresh, str) and new_refresh:
            self._refresh_token = new_refresh
        self._persist_tokens(token_response)
        self._logger.debug("PKCE token refreshed silently")
        return token_response

    def _build_token_response(self, parsed: dict[str, Any]) -> AccessTokenResponse:
        expires_in = parsed.get("expires_in")
        expires_in = expires_in if isinstance(expires_in, (int, float)) else 0
        expires = _now() + timedelta(seconds=expires_in)
        scope = parsed.get("scope")
        scopes = scope.split(" ") if isinstance(scope, str) else (self._config.scopes or [])
        return AccessTokenResponse(access_token=parsed["access_token"], expires=expires, scopes=scopes)

    def _persist_tokens(self, token_response: AccessTokenResponse) -> None:
        if not self._persist_session:
            return
        sub = self._sub
        try:
            decoded = decode_jwt(token_response.access_token)
            payload_sub = decoded.payload.get("sub")
            if isinstance(payload_sub, str) and payload_sub:
                sub = payload_sub
        except Exception:  # noqa: BLE001 - token may not be a JWT
            pass
        self._sub = sub
        record = AuthSession(
            client_id=self._config.client_id,
            flow="pkce",
            access_token=token_response.access_token,
            refresh_token=self._refresh_token,
            sub=sub,
            expires_at=_to_iso(token_response.expires),
            scopes=token_response.scopes,
            account_manager_host=self._account_manager_host,
        )
        try:
            save_auth_session(record)
        except Exception as error:  # noqa: BLE001 - persistence is best-effort
            self._logger.debug("Failed to persist PKCE session: %s", error)

    async def _run_flow(self) -> AccessTokenResponse:
        verifier, challenge = _generate_pkce_pair()
        state = _base64url(secrets.token_bytes(16))

        params = {
            "client_id": self._config.client_id,
            "redirect_uri": self._redirect_uri,
            "response_type": "code",
            "code_challenge": challenge,
            "code_challenge_method": "S256",
            "state": state,
        }
        if self._config.scopes:
            params["scope"] = " ".join(self._config.scopes)
        authorize_url = f"https://{self._account_manager_host}/dwsso/oauth2/authorize?{_form_encode(params)}"

        self._logger.info("Login URL: %s", authorize_url)
        self._logger.info("If the URL does not open automatically, copy/paste it into a browser on this machine.")

        code = await self._wait_for_auth_code(state, authorize_url)
        self._logger.debug("Got authorization code, exchanging for token")

        token_url = f"https://{self._account_manager_host}/dwsso/oauth2/access_token"
        token_body = _form_encode(
            {
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": self._redirect_uri,
                "client_id": self._config.client_id,
                "code_verifier": verifier,
            }
        )
        token_res = await dispatch_fetch(
            token_url,
            method="POST",
            headers={"Content-Type": "application/x-www-form-urlencoded"},
            content=token_body,
        )
        raw_text = token_res.text
        if not token_res.is_success:
            oauth_error, description = _parse_oauth_error_body(raw_text)
            detail = description or oauth_error
            message = f"PKCE token exchange failed ({token_res.status_code})" + (f": {detail}" if detail else "")
            if _is_pkce_grant_unsupported_error(oauth_error):
                raise PkceGrantUnsupportedError(message, "token", oauth_error)
            raise RuntimeError(message)

        try:
            parsed = json.loads(raw_text)
        except Exception as error:  # noqa: BLE001 - do not echo an untrusted body
            raise RuntimeError("PKCE token exchange returned a non-JSON response") from error

        access_token = parsed.get("access_token")
        if not isinstance(access_token, str) or not access_token:
            raise RuntimeError("PKCE token exchange response did not contain an access token")

        token_response = self._build_token_response(parsed)
        refresh = parsed.get("refresh_token")
        if isinstance(refresh, str) and refresh:
            self._refresh_token = refresh
        self._persist_tokens(token_response)
        return token_response

    async def _wait_for_auth_code(self, expected_state: str, authorize_url: str) -> str:
        opener = self._config.open_browser or open_browser_default

        def _handler(path: str, query: dict[str, list[str]]) -> CallbackResponse:
            code = _first(query, "code")
            state = _first(query, "state") or ""
            error = _first(query, "error")
            error_description = _first(query, "error_description")

            if error:
                message = f"OAuth error: {error_description or error}"
                auth_error: Exception = (
                    PkceGrantUnsupportedError(message, "authorize", error)
                    if _is_pkce_grant_unsupported_error(error)
                    else RuntimeError(message)
                )
                return CallbackResponse(
                    500,
                    "text/plain",
                    f"Authentication failed: {error_description or error}",
                    done=True,
                    error=auth_error,
                )
            if not code:
                return CallbackResponse(404, "text/plain", "Waiting for authorization code...")
            if state != expected_state:
                return CallbackResponse(
                    400,
                    "text/plain",
                    "State mismatch.",
                    done=True,
                    error=RuntimeError("OAuth state mismatch — aborting"),
                )
            return CallbackResponse(
                200,
                "text/plain",
                "Authorization received. Completing authentication in the application...",
                result=code,
                done=True,
            )

        self._logger.info("Waiting for user to authenticate...")

        async def _on_listening() -> None:
            await opener(authorize_url)

        result = await run_callback_server(self._local_port, _handler, _on_listening)
        return str(result)


def _parse_int(value: str | None) -> int | None:
    if not value:
        return None
    try:
        return int(value)
    except ValueError:
        return None


def _first(query: dict[str, list[str]], key: str) -> str | None:
    values = query.get(key)
    return values[0] if values else None


def _now() -> datetime:
    return datetime.now(tz=timezone.utc)


def _to_iso(value: datetime) -> str:
    return value.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


def _parse_iso(value: str) -> datetime:
    normalized = value.replace("Z", "+00:00")
    parsed = datetime.fromisoformat(normalized)
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed


def _form_encode(params: dict[str, str]) -> str:
    return urlencode(params)


__all__ = [
    "PkceOAuthConfig",
    "PkceOAuthStrategy",
    "PkceGrantUnsupportedError",
    "reset_pkce_cache_for_testing",
]
